apps/test-app/frontend/src/components/tree-widget/stateless-v2/Tree.tsx

```diff
@@ -0,0 +1,227 @@
+import {
+  createLimitingECSqlQueryExecutor, HierarchyProvider, ILimitingECSqlQueryExecutor, IMetadataProvider, TypedPrimitiveValue,
+} from "@itwin/presentation-hierarchy-builder";
+import { ModelsTreeDefinition } from "@itwin/presentation-models-tree";
+import { isPresentationHierarchyNode, PresentationTreeNode } from "./Types";
+import { UseTreeResult, useUnifiedSelectionTree } from "./UseTree";
+
+interface MetadataProviders {
+  queryExecutor: ILimitingECSqlQueryExecutor;
+  metadataProvider: IMetadataProvider;
+}
+
+export function StatelessTreeV2({ imodel, height, width }: { imodel: IModelConnection; height: number; width: number }) {
+  const [metadata, setMetadata] = useState<MetadataProviders>();
+  const [hierarchyProvider, setHierarchyProvider] = useState<HierarchyProvider>();
+  const [filter, setFilter] = useState("");
+  const [isFiltering, setIsFiltering] = useState(false);
+
+  useEffect(() => {
+    const schemas = new SchemaContext();
+    schemas.addLocater(new ECSchemaRpcLocater(imodel.getRpcProps()));
+    setMetadata({
+      queryExecutor: createLimitingECSqlQueryExecutor(createECSqlQueryExecutor(imodel), 1000),
+      metadataProvider: createMetadataProvider(schemas),
+    });
+  }, [imodel]);
+
+  const { value: filteredPaths } = useDebouncedAsyncValue(
+    useCallback(async () => {
+      setIsFiltering(false);
+      if (!metadata) {
+        return undefined;
+      }
+      if (filter !== "") {
+        setIsFiltering(true);
+        const paths = await ModelsTreeDefinition.createInstanceKeyPaths({
+          metadataProvider: metadata.metadataProvider,
+          queryExecutor: metadata.queryExecutor,
+          label: filter,
+        });
+        return paths;
+      }
+      return undefined;
+    }, [metadata, filter]),
+  );
+
+  useEffect(() => {
+    setIsFiltering(false);
+    if (!metadata?.metadataProvider) {
+      return;
+    }
+
+    setHierarchyProvider(
+      new HierarchyProvider({
+        metadataProvider: metadata.metadataProvider,
+        queryExecutor: metadata.queryExecutor,
+        hierarchyDefinition: new ModelsTreeDefinition({ metadataProvider: metadata?.metadataProvider }),
+        filtering: filteredPaths
+          ? {
+              paths: filteredPaths,
+            }
+          : undefined,
+      }),
+    );
+  }, [metadata, filteredPaths]);
+
+  const { rootNodes, isLoading, ...treeProps } = useUnifiedSelectionTree({
+    hierarchyProvider,
+  });
+
+  const [shouldUseCustomFormatter, setShouldUseCustomFormatter] = useState<boolean>(false);
+  const toggleFormatter = () => {
+    if (!hierarchyProvider) {
+      return;
+    }
+    const newValue = !shouldUseCustomFormatter;
+    hierarchyProvider.setFormatter(newValue ? customFormatter : undefined);
+    setShouldUseCustomFormatter(newValue);
+    treeProps.reloadTree();
+  };
+
+  const renderContent = () => {
+    if (rootNodes === undefined || isLoading || isFiltering) {
```

Comment:
On initial render we return `rootNodes` as `undefined` and `isLoading` as `false` because loading is not started yet. Also we return valid `rootNodes` array and `isLoading: true` when reloading tree so consumer could render old state until tree is reloaded but also indicate that reload is in progress.

Reply:
okay, please document that on `UseTreeResult`.